iroh-gateway/src/bad_bits.rs

```diff
@@ -0,0 +1,278 @@
+use cid::Cid;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+use std::{collections::HashSet, str::FromStr, sync::Arc, time::Duration};
+use tokio::{sync::RwLock, task::JoinHandle};
+use tracing::log::error;
+
+const BAD_BITS_UPDATE_INTERVAL: Duration = Duration::from_secs(3600 * 8);
+const DEFAULT_DENY_LIST_URI: &str = "http://badbits.dwebops.pub/denylist.json";
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub struct BadBitsAnchor {
+    pub anchor: String,
+}
+
+#[derive(Debug)]
+pub struct BadBits {
+    pub last_updated: time::Instant,
+    pub denylist: HashSet<String>,
+}
+
+impl BadBits {
+    pub fn new() -> Self {
+        Self {
+            last_updated: time::Instant::now(),
+            denylist: HashSet::new(),
+        }
+    }
+
+    pub fn update(&mut self, denylist: HashSet<String>) {
+        self.last_updated = time::Instant::now();
+        self.denylist = denylist;
+    }
+
+    pub fn is_bad(&self, cid: &str, path: &str) -> bool {
+        let cid = match Cid::from_str(cid) {
+            Ok(cid) => cid,
+            Err(_) => return false,
+        };
+        let hash = BadBits::to_anchor(cid, path);
+        self.denylist.contains(&hash)
+    }
+
+    pub fn to_anchor(cid: Cid, path: &str) -> String {
+        let mut hasher = Sha256::new();
+        hasher.update(cid.into_v1().unwrap().to_string());
+        if !path.is_empty() {
+            if path.starts_with('/') {
+                hasher.update(path);
+            } else {
+                hasher.update("/");
+                hasher.update(path);
+            }
+        } else {
+            hasher.update("/");
+        }
+        let x = hasher.finalize().to_vec();
+        let mut s = String::new();
```

Comment:
why store the hex in memory, could just store the hash and deserialize that when receiving the deny list. this is pretty expensive given how often we check it